R/animint.R

```diff
@@ -22,11 +22,54 @@ parsePlot <- function(meta){
       plot.meta$scales[[sc$aesthetics]] <- makeScale(sc)
     }
   }
+  
+  ## Export axis specification as a combination of breaks and
+  ## labels, on the relevant axis scale (i.e. so that it can
+  ## be passed into d3 on the x axis scale instead of on the
+  ## grid 0-1 scale). This allows transformations to be used
+  ## out of the box, with no additional d3 coding.
+  theme.pars <- ggplot2:::plot_theme(meta$plot)
+  
+  ## No legend if theme(legend.postion="none").
+  plot.meta$legend <- if(theme.pars$legend.position != "none"){
+    getLegendList(meta$built)
+  }
+  
+  ## save each layer
   for(layer.i in seq_along(meta$plot$layers)){
     ##cat(sprintf("%4d / %4d layers\n", layer.i, length(meta$plot$layers)))
 
     ## This is the layer from the original ggplot object.
     L <- meta$plot$layers[[layer.i]]
+    
+    ## If any legends are specified, add showSelected aesthetic
+    for(legend.i in seq_along(plot.meta$legend)) {
+      # the type of legend: colour, fill, etc
+      legend_type <- plot.meta$legend[[legend.i]]$legend_type
+      # the name of the variable used in this legend
+      var_name <- plot.meta$legend[[legend.i]]$vars
+      # the actual values for that variable
+      var <- L$data[[var_name]]
+      # checking if it is a discrete variable
+      if(plyr::is.discrete(var)) {
+        # adding a showSelected aesthetic for each legend entry
+        for(i in legend_type) {
+          temp_name <- paste0("showSelected", i)
+          L$mapping[[temp_name]] <- as.symbol(var_name)
+#           browser()
+        }
+        # if first is not specified, add all to first
+        if(is.null(meta$first[[var_name]])) {
+          meta$first[[var_name]] <- unique(var)
+        }
+        # if selectortypes is not specified, set it to multiple
+        if(is.null(meta$selector.types[[var_name]])) {
+          meta$selector.types[[var_name]] <- "multiple"
+        }
+      }
+    }
+    ## need to call ggplot_build again because I've added to the plot
+    meta$built <- ggplot2::ggplot_build(meta$plot)
```

Comment:
Please make it as detailed as possible. I would recommend just copy-pasting your comment from above:

```As for calling ggplot_build twice, I'm sure that there is a way around it, but I'm not immediately sure how. There's sort of a Catch-22 here because to create the interactivity, we need to specify the variable corresponding to each legend. To do this, we need to have the legend. And to have the legend, I think that we need to use ggplot_build. Does that make sense? Let me know if not.
```

More generally if you have to make some design decisions it is better to provide justification in the code comments than in the github comments, so it is easier to remember why we made those decisions when we have to come back to the code later.